Remove actor table info from storage when a driver exits

WangTaoTheTonic opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

When mark a job as finished, we just change the status in job table. We need also clean the actor related information(actor/checkpoint/checkpoint id table) stored in gcs storage too

1. only non-detached actor should be cleaned
2. object informations is cleaned by raylet by reference couting
3. task will not be stored in gcs storage any more in future

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26695/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26699/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26777/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26836/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26853/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26912/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/26934/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27019/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27018/
Test FAILed.

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_actor_manager.cc:
```diff
@@ -764,5 +764,77 @@ void GcsActorManager::LoadInitialData(const EmptyCallback &done) {
+}
+
+void GcsActorManager::ClearActorStates(std::vector<ActorID> actor_ids) {
+  for (auto &actor_id : actor_ids) {
```
This logic looks like it duplicates the `DestroyActor` logic. Please use that method instead.

Also, I don't think that we actually need to call this method when the job finishes. If the driver exits, all actors related to the driver will eventually exit via fate-sharing. Once the actor's owner exits, the GCS actor manager will call `DestroyActor` anyway. Perhaps we should modify `DestroyActor` to also delete the Redis entry (which I notice you're not doing here, any reason for that?), but that seems low-priority right now, and we should wait for #8679 to be merged before we do that.

WangTaoTheTonic replied:
At first I tried to reuse `DestroyActor` but it seems like will do some checks while what we need is to clear in-memory state.
After job finished we need to clear in-memory state and the entries(actor/checkpoint/checkpoint id table) in Redis of its actors. These two steps are now done in `OnJobFinished`.
As you said if all in-memory state will be cleared after driver exists, maybe we just delete Redis entry in `OnJobFinished` method.

stephanie-wang replied:
Yes, I think that makes sense (just deleting the Redis entry in OnJobFinished).

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_actor_manager.h:
```diff
@@ -223,6 +223,10 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   /// \param done Callback that will be called when load is complete.
   void LoadInitialData(const EmptyCallback &done);
 
+  void OnJobFinished(const JobID &job_id);
```
Documentation.

WangTaoTheTonic replied:
ok

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_object_manager.cc:
```diff
@@ -288,6 +315,12 @@ void GcsObjectManager::LoadInitialData(const EmptyCallback &done) {
 
+void GcsObjectManager::OnJobFinished(const JobID &job_id) {
+  /// Remove caches in object manager first, then items in object table.
+  RemoveAllObjectLocationsInCacheByJobId(job_id);
```
This shouldn't be necessary. The ref counting will ensure that all raylets will remove old objects. This is also pretty sketchy since the GCS will become out of sync with the plasma stores.

WangTaoTheTonic replied:
I found the caches will not be remove immediately as test here https://github.com/ray-project/ray/pull/8761/files#diff-b13b9c2b24b42796bde800d10df4f030R603 will fail if we leave the old object alone.

Will dig more to confirm.

stephanie-wang replied:
Hmm but that test is not a full integration test. In the e2e system, the object table will eventually get cleaned up by the raylets.

Actually, one piece of state that does not currently get removed is if the raylet dies. It makes sense for the GCS to clean up the objects that were on the dead raylet.

stephanie-wang commented on src/ray/gcs/gcs_server/task_info_handler_impl.cc:
```diff
@@ -173,5 +173,12 @@ void DefaultTaskInfoHandler::HandleAttemptTaskReconstruction(
 }
 
+void DefaultTaskInfoHandler::OnJobFinished(const JobID &job_id) {
+  RAY_CHECK_OK(gcs_table_storage_->TaskTable().DeleteByJobId(job_id, nullptr));
```
This shouldn't be necessary. The new architecture and GCS-based actor management codepath never writes to the task table.

WangTaoTheTonic replied:
ok. Heard that from @raulchen

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27070/
Test PASSed.

stephanie-wang commented:
Okay, so we should narrow the scope of this PR to just cleaning up the actor checkpoint table and actor table entries in Redis, right? I think the rest of the state should already be taken care of, but it'd be good to confirm that with a Python integration test.

You could also consider addressing the comment that I left about cleaning up objects that are on dead nodes, but it's not high priority.

Another thing that I think would make sense is cleaning up the function table after a job exits. Right now, this is set directly in Redis through Python ([code](https://github.com/ray-project/ray/blob/master/python/ray/function_manager.py#L142)).

WangTaoTheTonic commented:
> Okay, so we should narrow the scope of this PR to just cleaning up the actor checkpoint table and actor table entries in Redis, right? I think the rest of the state should already be taken care of, but it'd be good to confirm that with a Python integration test.
> 
> You could also consider addressing the comment that I left about cleaning up objects that are on dead nodes, but it's not high priority.
> 
> Another thing that I think would make sense is cleaning up the function table after a job exits. Right now, this is set directly in Redis through Python ([code](https://github.com/ray-project/ray/blob/master/python/ray/function_manager.py#L142)).

Got it. I've reviewed all your comments and appreciate them very much. Garbage collection of a job means clear everything about it once a job finished, but as you said simply remove them in one place is lack of consideration.
As you suggest, we can clear the things that we are sure. About the unsure part(or parts should be co-work with other components) we can fix them step by step in future.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27150/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27156/
Test PASSed.

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_actor_manager.h:
```diff
@@ -223,6 +223,11 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   /// \param done Callback that will be called when load is complete.
   void LoadInitialData(const EmptyCallback &done);
 
+  /// Clear related actor information once the job it belongs finished.
```
```suggestion
  /// Delete non-detached actor information from durable storage once the associated job finishes.
```

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_object_manager.h:
```diff
@@ -97,6 +97,9 @@ class GcsObjectManager : public rpc::ObjectInfoHandler {
       LOCKS_EXCLUDED(mutex_);
 
+  /// Remove all cached object locations for given job id.
+  void RemoveAllObjectLocationsInCacheByJobId(const JobID &job_id);
```
We should remove this, right?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/27196/
Test FAILed.

WangTaoTheTonic commented:
![image](https://user-images.githubusercontent.com/5276001/84726553-b811a580-afbf-11ea-83eb-efe94775669c.png)
Seems like the disk is full.... (https://amplab.cs.berkeley.edu/jenkins/job/Ray-PRB/27196/console)
